File: Algorithms_4th_Edition/c/1/2/book/counter.c
#include <stdio.h>
#include "counter.h"

/**
 * 初始化一个名为id的计数器
 * param: c - 计数器存储的数据对象
 * param: id - 初始化计数器名字的
 */ 
void initCounter(Counter *c,char * id)
{
    c->id = id;
    c->data = 0;
}

/**
 * 将计数器的值加1
 * param: c是计数器存储的数据指针
 */ 
void increment(Counter *c)
{
    c->data++;
}

/**
 * 该对象创建之后计算器被加1的次数
 * param: c是计数器存储的数据指针
 * return: 返回被加1的次数
 */
int tally(const Counter *c)
{
    return c->data;
}

void counterString(const Counter * c,char * st,int n)
{
    snprintf(st,n,"%d %s",c->data,c->id);
}

File: Algorithms_4th_Edition/java/src/1/2/book/Counter.java
public class Counter {

    private final String name;
    private int count;

    /**
     * 创建一个名为id的计数器
     * @param id 计数的名
     */
    public Counter(String id) {
        this.name = id;
    }

    /**
     * 将计数器的值加一
     */
    public void increment(){
        this.count++;
    }

    public int tally(){
        return this.count;
    }

    @Override
    public String toString() {
        return count + " " + name;
    }

    public static void main(String[] args) {
        Counter heads = new Counter("heads");
        Counter tails = new Counter("tails");

        heads.increment();
        heads.increment();
        tails.increment();

        System.out.println(heads + " " + tails);
        System.out.println(heads.tally() + tails.tally());
    }
}
